lib/matplotlib/colors.py

```diff
@@ -1099,14 +1099,37 @@ def from_list(name, colors, N=256, gamma=1.0):
         if not np.iterable(colors):
             raise ValueError('colors must be iterable')
 
-        if (isinstance(colors[0], Sized) and len(colors[0]) == 2
-                and not isinstance(colors[0], str)):
-            # List of value, color pairs
-            vals, colors = zip(*colors)
+        try:
+            # Assume the passed colors are a list of colors
+            # and not a (value, color) tuple.
+            r, g, b, a = to_rgba_array(colors).T
+        except (ValueError, TypeError) as e:
+            # Check whether the first element is a (value, color) tuple
+            if (
```

Comment:
rather than try to enumerate the ways this could work I would do it something like

```
try:
    _val, _colors = itertools.zip_longest(*colors)
except Exception as e2:
    # or some type checking / work to make a better exception, but show both
    # as we don't know which the user meant!
    raise e2 from e
val = np.asarray(_val, dytpe=float)
if val.min() < 0 or val.max() > 1 or np.any(np.diff(val) <= 0):
    # TODO better message
    raise ValueError("broken values")
r, g, b, a = to_rgba_array(_colors).T
```

This way the "happy path" of things working is to the left as possible, the (some what messy) work of sorting out how to write a good error message is contained in the `except` blocks, and avoid we rely as much as possible on duck-typing (if it works it must be the right type!) rather than trying to enumerate types. 

----

This big `if` is still only checking the first entry in the input rather than all of them.

Reply:
Totally agree! I like this code a lot more, looks much cleaner. I will incorporate your suggestions in the next revision ;)